src/array_api_extra/_funcs.py

```
@@ -548,3 +559,279 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+
+    def __init__(self, x: Array, idx: Index = _undef, /):
+        self.x = x
+        self.idx = idx
+
+    def __getitem__(self, idx: Index) -> at:
+        """Allow for the alternate syntax ``at(x)[start:stop:step]``,
+        which looks prettier than ``at(x, slice(start, stop, step))``
+        and feels more intuitive coming from the JAX documentation.
+        """
+        if self.idx is not _undef:
+            msg = "Index has already been set"
+            raise ValueError(msg)
+        self.idx = idx
+        return self
+
+    def _common(
+        self,
+        at_op: str,
+        y: Array = _undef,
+        /,
+        copy: bool | None = True,
+        xp: ModuleType | None = None,
+        _is_update: bool = True,
+        **kwargs: Untyped,
+    ) -> tuple[Untyped, None] | tuple[None, Array]:
+        """Perform common prepocessing.
+
+        Returns
+        -------
+        If the operation can be resolved by at[], (return value, None)
+        Otherwise, (None, preprocessed x)
+        """
+        if self.idx is _undef:
+            msg = (
+                "Index has not been set.\n"
+                "Usage: either\n"
+                "    at(x, idx).set(value)\n"
+                "or\n"
+                "    at(x)[idx].set(value)\n"
+                "(same for all other methods)."
+            )
+            raise TypeError(msg)
+
+        x = self.x
+
+        if copy is True:
+            writeable = None
+        elif copy is False:
+            writeable = is_writeable_array(x)
+            if not writeable:
+                msg = "Cannot modify parameter in place"
+                raise ValueError(msg)
+        elif copy is None:  # type: ignore[redundant-expr]
+            writeable = is_writeable_array(x)
+            copy = _is_update and not writeable
+        else:
+            msg = f"Invalid value for copy: {copy!r}"  # type: ignore[unreachable]  # pyright: ignore[reportUnreachable]
```

is there a better way to deal with situations like this where invalid types can be passed at runtime?

---

This double-ignore is a strong example of why I'm strongly in favour of having only one type checker. This is tedious to both read and write.

---

it's definitely annoying, but unfortunately it's the only possibility for ensuring compatibility with these two main type-checkers

---

but I agree with mypy and pyright that the `else` clause here is redundant

---

you could alternatively move the `if copy is None` to the top, and then do `elif copy: ... else: ...`

---

what if we still want to throw an error at runtime if `copy="foo"` is passed? I guess the argument is that, since this is internal, that would always be caught by the static analysis anyway?